Sources/Basic/OutputByteStream.swift

```diff
@@ -55,27 +60,89 @@ public class OutputByteStream: TextOutputStream {
-    
+
+     /// Clears the buffer maintaining current capacity.
+    private func clearBuffer() {
+        buffer.removeAll(keepingCapacity: true)
+    }
+
     // MARK: Data Output API
 
     public final func flush() {
+        writeImpl(buffer)
+        clearBuffer()
+        flushImpl()
+    }
+
+    func flushImpl() {
         // Do nothing.
     }
 
+    func writeImpl<C: Collection where C.Iterator.Element == UInt8>(_ bytes: C) {
+        fatalError("Subclasses must implement this")
+    }
+
     /// Write an individual byte to the buffer.
     public final func write(_ byte: UInt8) {
+        // If buffer is full, write and clear it.
+        if availableBufferSize == 0 {
+            writeImpl(buffer)
+            clearBuffer()
+        }
+
+        // This will need to change change if we ever have unbuffered stream.
+        precondition(availableBufferSize > 0)
         buffer.append(byte)
     }
 
     /// Write a collection of bytes to the buffer.
-    public final func write<C: Collection where C.Iterator.Element == UInt8>(collection bytes: C) {
+    public final func write<
+                            C: Collection where C.Iterator.Element == UInt8,
+                            C.IndexDistance == Int,
+                            C.SubSequence: Collection,
+                            C.SubSequence.Iterator.Element == UInt8
+                            >(collection bytes: C) {
+        // This is based on LLVM's raw_ostream.
+        let availableBufferSize = self.availableBufferSize
+
+        // If we have to insert more than the available space in buffer.
+        if bytes.count > availableBufferSize {
+            // If buffer is empty, start writing and keep the last chunk in buffer.
+            if buffer.isEmpty {
+                let bytesToWrite = bytes.count - (bytes.count % availableBufferSize)
+                let writeUptoIndex = bytes.index(bytes.startIndex, offsetBy: bytesToWrite)
+                writeImpl(bytes.prefix(upTo: writeUptoIndex))
+
+                // If remaining bytes is more than buffer size write everything.
+                let bytesRemaining = bytes.count - bytesToWrite
+                if bytesRemaining > availableBufferSize {
+                    writeImpl(bytes.suffix(from: writeUptoIndex))
+                    return
+                }
+                // Otherwise keep remaining in buffer.
+                buffer += bytes.suffix(from: writeUptoIndex)
+                return
+            }
+
+            let writeUptoIndex = bytes.index(bytes.startIndex, offsetBy: availableBufferSize)
+            // Append whatever we can accomodate.
+            buffer += bytes.prefix(upTo: writeUptoIndex)
+
+            writeImpl(buffer)
+            clearBuffer()
+
+            // FIXME: We should start again with remaining chunk but this doesn't work. Write everything for now.
+            //write(collection: bytes.suffix(from: writeUptoIndex))
+            writeImpl(bytes.suffix(from: writeUptoIndex))
```

Comment:
The compiler is lying here, a good trick is to split the expression into temporaries. The actual problem is that it can't infer `C` in the write call from a `C.SubSequence`.